Change ConfigurationBinder to avoid reading properties unnecessarily.

madelson opened the pull request:
Previously, ConfigurationBinder would eagerly read properties it
encountered and initialize/bind objects for them even when those
reads/initializations could lead to thrown exceptions or unnecessary
work. Now, we defer work until we determine that it will actually
contribute to the final binding result.

Fix #52905
Fix #65710

madelson committed:
Change ConfigurationBinder to avoid reading properties unnecessarily.

Previously, ConfigurationBinder would eagerly read properties it
encountered and initialize/bind objects for them even when those
reads/initializations could lead to thrown exceptions or unnecessary
work. Now, we defer work until we determine that it will actually
contribute to the final binding result.

Fix #52905
Fix #65710

madelson commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -233,29 +234,27 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
         {
             Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments[0]);
-            object? instance = Activator.CreateInstance(genericType);
+            object instance = Activator.CreateInstance(genericType)!;
```
Calling Activator.CreateInstance on a List can't ever return null, right?

tarekgh replied:
It is unlikely to return null for List<T>

madelson replied:
That's what I thought; just wanted to highlight the change.

madelson committed:
Restore support for binding mutable structs.

madelson commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -415,19 +417,28 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
             }
+
+            if (instance is null
+                && Nullable.GetUnderlyingType(type) is { } nullableUnderlyingType)
```
This is technically new functionality. Previously, null mutable structs that were initially null would end up nooping in BindNonScalar. In my mind it makes more sense to treat these the same whether they start as null or not.

halter73 replied:
This makes sense to me, but I think we should wrap any exceptions thrown from `Activator.CreateInstance(nullableUnderlyingType)`.

```C#
            try
            {
                if (Nullable.GetUnderlyingType(type) is { } nullableUnderlyingType)
                {
                    // Activator.CreateInstance only returns null for Nullable<T> types.
                    return Activator.CreateInstance(nullableUnderlyingType)!;
                }

                return Activator.CreateInstance(type)!;
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException(SR.Format(SR.Error_FailedToActivate, type), ex);
            }
```

If we don't believe `Activator.CreateInstance only returns null for Nullable<T> types`, we should be checking the return values for both calls to `Activator.CreateInstance` to see if they're null.

madelson replied:
>  I think we should wrap any exceptions thrown from Activator.CreateInstance(nullableUnderlyingType).

Good call. I forgot that this can happen with default struct constructors.

> If we don't believe Activator.CreateInstance only returns null for Nullable<T> types, we should be checking the return values for both calls to Activator.CreateInstance to see if they're null.

@halter73 I'll defer to you on this. I vaguely recall reading a C# article years ago that said some COM objects could result in `null` returns from `Activator.CreateInstance`. No idea whether that's still true in .NET Core, whether we care if it is, or whether there are other edge cases like that to consider. Unfortunately, I can't find the original source. If we think explicitly handling `Nullable<T>` is sufficient to guarantee a non-null return, we can just use `!` as in your example.

tarekgh replied:
@madelson I tried to look if `Activator.CreateInstance` can return null for other cases but couldn't know if this is the case. Should we do the null check anyway to be in the safe side?

@halter73 what you think?

halter73 replied:
> Should we do the null check anyway to be in the safe side?

I'm also unsure if `Activator.CreateInstance` can return null in other cases. It's the only scenario I see where [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.activator.createinstance?view=net-6.0#system-activator-createinstance(system-type)) say `CreateInstance` will return null, but I think null checking all created instances as the PR currently does (except for with well-known types like `List<T>`) makes sense.

madelson replied:
Thanks! Anything else for me to do here @tarekgh @halter73 ?

madelson committed:
Attempt to appease IL2072

madelson committed:
Merge remote-tracking branch 'origin/main' into issue-52905

# Conflicts:
#	src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

madelson committed:
Add RequiresUnreferencedCode to CreateInstance

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:
```diff
@@ -1043,6 +1043,80 @@ public void ExceptionWhenTryingToBindToByteArray()
                 exception.Message);
         }
 
+        [Fact]
```
>Fact [](http://example.com/codeflow?start=9&length=4)

do we have a test case like the case mentioned in [https://github.com/lenraven/BindPropertyExceptionSample/tree/master/BindPropertyExceptionSample](https://github.com/lenraven/BindPropertyExceptionSample/tree/master/BindPropertyExceptionSample)?

madelson replied:
I tried to test that with https://github.com/dotnet/runtime/pull/66723/files#diff-22b129d6955920524411aeb76c26d711b280b02f9b44bf043c0b9ffa22aed564R1166 . Always throwing is the "harder" case compared to conditionally throwing; in both cases the key is not to read the property value if we don't need to.

If you wanted, I could add a case that is specifically property order dependent. Something like this:
```
class OrderMatters
{
    private readonly List<string> _list = new();

    public string? BindThisFirst { get; set; }

    public List<string> BindThisSecond => BindThisFirst is null ? throw new InvalidOperationException() : _list;
}
```

Do you think that would be valuable? Is property reflection order even well-defined in .NET Core (it wasn't in .NET Framework).

tarekgh replied:
I was hoping to test the case like 

```C#
public class Configuration
{
    public IEnumerable<int> FilteredConfigValues => ConfigValues.Where(p => p > 10);
}   public IEnumerable<int> ConfigValues { get; set; } = null!;
 ```

today this throws because `FilteredConfigValues` is initialized before `ConfigValues `. if the order is flipped like the following, it will not throw any more. 

```C#
public class Configuration
{
    public IEnumerable<int> ConfigValues { get; set; } = null!;
    public IEnumerable<int> FilteredConfigValues => ConfigValues.Where(p => p > 10);
}
```

Adding such a test case will validate the fix we are trying to do here.

madelson replied:
Added!

maryamariyan commented:
Thanks for the PR @madelson, the changes at first glance seem fine, thanks the test cases look good too.

Note for reviewers,
This issue had been reported earlier too and we reverted its fix since it caused a regression.
Link to the PR reverting original fix: https://github.com/dotnet/runtime/pull/63586
Let's make sure to add a test case for the regression scenario that was reported in https://github.com/dotnet/runtime/issues/63479 (Seems like a similar use case is already included).

/cc @halter73 would be good to get your review on this as well.

tarekgh commented:
Just to note, if flipping the order of the properties, this should workaround this issue:

```C#
public class Configuration
{
    public IEnumerable<int> ConfigValues { get; set; } = null!;
    public IEnumerable<int> FilteredConfigValues => ConfigValues.Where(p => p > 10);
}
```

madelson committed:
Add additional test coverage for ConfigurationBinder.

Address feedback from
https://github.com/dotnet/runtime/pull/66723#discussion_r835645244 and
from https://github.com/dotnet/runtime/pull/66723#issuecomment-1079411922.

madelson commented:
> Let's make sure to add a test case for the regression scenario that was reported in https://github.com/dotnet/runtime/issues/63479 (Seems like a similar use case is already included).

@maryamariyan I added a new test case explicitly reproducing this issue.

tarekgh reviewed (approved):
LGTM. Thanks @madelson!

@halter73 @davidfowl  are you interested to take a look at this change?

halter73 commented on src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:
```diff
@@ -344,77 +343,72 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
             if (config != null && config.GetChildren().Any())
             {
+                // for arrays and read-only list-like interfaces, we concatenate on to what is already there
+                if (type.IsArray || IsArrayCompatibleReadOnlyInterface(type))
```
Prior to this change, we'd assign `List<T>`s to settable `IEnumerable<T>`, `IReadOnlyCollection<T>` and `IReadOnlyList<T>` properties. Now these will be arrays. I think this is okay, just pointing it out.

madelson replied:
@halter73 good callout; I should have left a comment about this. I made the change just to reduce some otherwise-duplicative code. As you say, I think it should be ok. Hopefully noone is assuming they can cast one of the read-only interfaces to a specific concrete implementation.

halter73 commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:
```diff
@@ -1043,6 +1043,80 @@ public void ExceptionWhenTryingToBindToByteArray()
+            ConfigurationBuilder configurationBuilder = new();
+            configurationBuilder.AddInMemoryCollection(new Dictionary<string, string>
+            {
+                { nameof(ClassWithReadOnlyPropertyThatThrows.Safe), "value" }
```
Can we verify these don't throw even if they're configured?

```suggestion
                { nameof(ClassWithReadOnlyPropertyThatThrows.Safe), "value" },
                { nameof(ClassWithReadOnlyPropertyThatThrows.StringThrows), "value" },
                { $"{nameof(ClassWithReadOnlyPropertyThatThrows.EnumerableThrows)}:0", "0" },
```

halter73 reviewed (approved):
LGTM aside from the two minor comments.

tarekgh commented:
@madelson could you please apply @halter73 suggestions to proceed here? Thanks!

madelson committed:
Update src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs

Review feedback from https://github.com/dotnet/runtime/pull/66723#discussion_r836957572

Co-authored-by: Stephen Halter <halter73@gmail.com>

madelson committed:
Wrap exceptions thrown by Activator.CreateInstance for nullable types.

Address feedback from
https://github.com/dotnet/runtime/pull/66723/files/e2049053a87206c6776c00741a4deee6585511c8#r836953237

madelson commented:
@tarekgh I've applied the changes. One very small question to be resolved here: https://github.com/dotnet/runtime/pull/66723#discussion_r838418283

tarekgh commented:
Thanks @madelson for helping with this PR.
